### PR TITLE
Fix AtomicLocalFile #2087 erroneous move (py2), #2599 dangling open file handle (win,nix?)

### PR DESCRIPTION
**Pull request by computercolin**
## Description
Fixes to AtomicLocalFile to fix 

## Motivation and Context
Bugs where AtomicLocalFile
  - erroneously commits failed data
  - fails to cleanup open file handles
  - raises "file open in other process" error and dies (on windows)

## Have you tested this? If so, how?
Only by hand. Tested with simple task pipe, and tested using reproduction code in #2087 and #2599

### CODE REVIEW

**Comment by dlstadther** [comment]
@computercolin are there any tests which should already be covering this patch that weren't?

File: luigi/target.py [hunk 1]
```diff
@@ -316,18 +316,23 @@ class AtomicLocalFile(io.BufferedWriter):
...
+    Subclasses must implement `move_move_to_final_destination`.
+
+    :class:`luigi.local_target.atomic_file` example of local-endpoint atomic-file.
+    :class:`luigi.contrib.s3.AtomicS3File` example of remote-endpoint usage.
     """
 
     def __init__(self, path):
         self.__tmp_path = self.generate_tmp_path(path)
         self.path = path
+        self.unintended_close = False
```
Comment (Tarrasch) [inline luigi/target.py]: Make this private?

File: luigi/target.py [hunk 2]
```diff
@@ -336,6 +341,8 @@ def move_to_final_destination(self):
         raise NotImplementedError()
 
     def __del__(self):
+        self.unintended_close = True
+        self.close()
```
Comment (Tarrasch) [inline luigi/target.py]: This pattern is hard to read I think. If `close()` depends on the attribute `unintended_close` it make much more sense for it to be a parameter to `close()`

File: luigi/target.py (hunk 1 above)
Comment (Tarrasch) [reply luigi/target.py]: Also this variable name is not descriptive I think. I don't understand what it does.

File: luigi/target.py [hunk 3]
```diff
@@ -316,18 +316,23 @@ class AtomicLocalFile(io.BufferedWriter):
     a temporary file in the local filesystem before
     moving it to its final destination.
 
-    This class is just for the writing part of the Target. See
-    :class:`luigi.file.LocalTarget` for example
+    This class handles only creation, cleanup of on-disk tmpfile.
+    Subclasses must implement `move_move_to_final_destination`.
```
Comment (Tarrasch) [inline luigi/target.py]: is `move_move_to_final_destination` correctly spelled?

**Pull request closed without merging**